File: poc/dds-reference/python/dds_client/dds_client/identity.py
"""DID -> PDS resolution. Mirror of `@dds/client`'s identity helper.

Supports `did:plc:*` (via plc.directory) and `did:web:*` (via the well-known DID document).
Results are cached in-process.
"""

from __future__ import annotations

import logging
from typing import Optional

import httpx

log = logging.getLogger(__name__)

_cache: dict[str, Optional[str]] = {}
DEFAULT_PLC_DIRECTORY = "https://plc.directory"


async def resolve_did_pds(
    did: str,
    *,
    plc_directory: str = DEFAULT_PLC_DIRECTORY,
) -> Optional[str]:
    """Return the `#atproto_pds` service endpoint URL for a DID, or None if not resolvable."""
    if did in _cache:
        return _cache[did]

    url = _did_document_url(did, plc_directory)
    if not url:
        _cache[did] = None
        return None

    try:
        async with httpx.AsyncClient(timeout=30.0) as http:
            resp = await http.get(url)
            resp.raise_for_status()
            doc = resp.json()
    except Exception as exc:
        log.warning("resolve_did_pds: failed for %s: %s", did, exc)
        _cache[did] = None
        return None

    endpoint = _extract_pds_endpoint(doc, did)
    _cache[did] = endpoint
    return endpoint


def clear_did_resolver_cache() -> None:
    """For testing: drop cached resolutions."""
    _cache.clear()


def _did_document_url(did: str, plc_directory: str) -> Optional[str]:
    if did.startswith("did:plc:"):
        return f"{plc_directory.rstrip('/')}/{did}"
    if did.startswith("did:web:"):
        rest = did[len("did:web:"):]
        segments = [_decode(s) for s in rest.split(":")]
        if not segments or not segments[0]:
            return None
        domain = segments[0]
        path = "/" + "/".join(segments[1:]) if len(segments) > 1 else ""
        return f"https://{domain}{path}/.well-known/did.json"
    return None


def _decode(s: str) -> str:
    from urllib.parse import unquote

    return unquote(s)


def _extract_pds_endpoint(doc: dict, did: str) -> Optional[str]:
    want_ids = {"#atproto_pds", f"{did}#atproto_pds"}
    for s in doc.get("service", []) or []:
        if not isinstance(s, dict):
            continue
        if s.get("id") in want_ids or s.get("type") == "AtprotoPersonalDataServer":
            ep = s.get("serviceEndpoint")
            if isinstance(ep, str):
                return ep
            if isinstance(ep, dict) and isinstance(ep.get("uri"), str):
                return ep["uri"]
    return None

File: poc/atproto/dds-reference/packages/client-ts/src/identity.ts
/**
 * DID -> PDS resolution. Used to route cross-repo reads to the correct PDS, since an
 * authenticated session is bound to its own PDS and can't fetch records hosted elsewhere.
 *
 * Supports `did:plc:*` (via plc.directory) and `did:web:*` (via the well-known DID document).
 * Results are cached in-process to avoid repeated lookups.
 */

const cache = new Map<string, string | null>();
let pending = new Map<string, Promise<string | null>>();

export interface ResolveDidOptions {
  /** Override the plc.directory base URL. */
  plcDirectory?: string;
}

/**
 * Resolve a DID to its `#atproto_pds` service endpoint URL, or `null` if the DID can't be
 * resolved (network failure, malformed doc, missing service, unsupported DID method).
 */
export async function resolveDidPds(
  did: string,
  options: ResolveDidOptions = {},
): Promise<string | null> {
  if (cache.has(did)) return cache.get(did)!;
  const existing = pending.get(did);
  if (existing) return existing;

  const p = (async () => {
    const url = didDocumentUrl(did, options);
    if (!url) return null;
    try {
      const res = await fetch(url);
      if (!res.ok) return null;
      const doc = (await res.json()) as DidDocument;
      const endpoint = extractPdsEndpoint(doc, did);
      cache.set(did, endpoint);
      return endpoint;
    } catch {
      cache.set(did, null);
      return null;
    } finally {
      pending.delete(did);
    }
  })();
  pending.set(did, p);
  return p;
}

/** For testing: drop cached resolutions. */
export function clearDidResolverCache(): void {
  cache.clear();
  pending = new Map();
}

interface DidDocument {
  service?: Array<{ id?: string; type?: string; serviceEndpoint?: string | { uri?: string } }>;
}

function didDocumentUrl(did: string, options: ResolveDidOptions): string | null {
  if (did.startsWith("did:plc:")) {
    const base = (options.plcDirectory ?? "https://plc.directory").replace(/\/$/, "");
    return `${base}/${did}`;
  }
  if (did.startsWith("did:web:")) {
    const rest = did.slice("did:web:".length);
    // did:web:domain[:path] -> https://domain[/path]/.well-known/did.json
    const segments = rest.split(":").map(decodeURIComponent);
    const domain = segments.shift();
    if (!domain) return null;
    const path = segments.length > 0 ? "/" + segments.join("/") : "";
    return `https://${domain}${path}/.well-known/did.json`;
  }
  return null;
}

function extractPdsEndpoint(doc: DidDocument, did: string): string | null {
  const services = doc.service ?? [];
  const want = new Set(["#atproto_pds", `${did}#atproto_pds`]);
  for (const s of services) {
    if (!s) continue;
    const matchesId = s.id && want.has(s.id);
    const matchesType = s.type === "AtprotoPersonalDataServer";
    if (!matchesId && !matchesType) continue;
    const ep = s.serviceEndpoint;
    if (typeof ep === "string") return ep;
    if (ep && typeof ep === "object" && typeof ep.uri === "string") return ep.uri;
  }
  return null;
}
